grin/src/Transformations/ExtendedSyntax/Optimising/CaseHoisting.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE LambdaCase, TupleSections #-}
+module Transformations.ExtendedSyntax.Optimising.CaseHoisting where
+
+import Control.Monad
+import Control.Comonad
+import Control.Comonad.Cofree
+import Data.Functor.Foldable as Foldable
+import qualified Data.Foldable
+
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Set (Set)
+import qualified Data.Set as Set
+import qualified Data.Vector as Vector
+import Data.Bifunctor (first)
+
+import Grin.ExtendedSyntax.Grin
+import Grin.ExtendedSyntax.TypeEnv
+import Transformations.ExtendedSyntax.Util
+import Transformations.ExtendedSyntax.Names
+
+{-
+  IDEA:
+    If Alt had name then the HPT could calculate it's return type and store in TypeEnv
```

Comment:
Alt has a name now. Maybe create an issue about it?

Reply:
Well, currently HPT tracks the alt name as a restricted version of the scrutinee. I will create an issue just for tracking this.

Reply:
See #94